src/time/time.h

```diff
@@ -1,95 +1,114 @@
- *   added in the future to strike different balance between precision and cost.
- */
-#include <cc_define.h>
-
-#include <inttypes.h>
+#include <stdint.h>
 #include <time.h>
 
-/* NOTE(yao): this whole time module needs a major overhaul */
+#include "util/define.h"
 
 /*
- * How time works internally for memcached
- *
- * Memcached server uses a timer with second granularity, which starts as the
- * process starts, and is set to 2 initially to avoid having 0 aged items.
- *
- * On systems where size(time_t) > sizeof(uint32_t), this gives
- * us space savings over tracking absolute unix time of type time_t
+ * For less granular time, rel_time_t will suffice and gives space savings.
+ * For more granular time, rel_time_fine_t gives additional precision.
  */
 typedef uint32_t rel_time_t;
+typedef uint64_t rel_time_fine_t;
 
 /*
- * From memcache protocol specification:
- *
- * Some commands involve a client sending some kind of expiration time
- * (relative to an item or to an operation requested by the client) to
- * the server. In all such cases, the actual value sent may either be
- * Unix time (number of seconds since January 1, 1970, as a 32-bit
- * value), or a number of seconds starting from current time. In the
- * latter case, this number of seconds may not exceed 60*60*24*30 (number
- * of seconds in 30 days); if the number sent by a client is larger than
- * that, the server will consider it to be real Unix time value rather
- * than an offset from current time.
+ * Time when the process was started expressed as absolute unix timestamp
  */
-#define TIME_MAXDELTA   (time_t)(60 * 60 * 24 * 30)
+extern time_t time_start;
 
 /*
- * Time when process was started expressed as absolute unix timestamp
- * with a time_t type
- */
-time_t time_start;
-
-/*
- * We keep a cache of the current time of day in a global variable now
- * that is updated periodically by a timer event every second. This
- * saves us a bunch of time() system calls because we really only need
- * to get the time once a second, whereas there can be tens of thosands
- * of requests a second.
- *
- * Also keeping track of time as relative to server-start timestamp
- * instead of absolute unix timestamps gives us a space savings on
- * systems where sizeof(time_t) > sizeof(unsigned int)
- *
- * So, now actually holds 32-bit seconds since the server start time.
+ * Current time relative to process start. These are updated with each call
+ * to time_update().
  */
-rel_time_t now;
+extern rel_time_t now_sec;
+extern rel_time_fine_t now_ms;
+extern rel_time_fine_t now_us;
+extern rel_time_fine_t now_ns;
 
-/* Get the time the process started */
+/* Get the time the process was started */
 static inline time_t
 time_started(void)
 {
-    return time_start;
+    return __atomic_load_n(&time_start, __ATOMIC_RELAXED);
+}
+
+/* Get the current time in seconds since the process started */
+static inline rel_time_t
+time_now(void)
```

Comment:
Do you think we can rename these primitives to make it more obvious when the clock started?
e.g., we can have:
`process_time_sec()`, `process_time_ms()` .... that were started when the process started.